### PR TITLE
Implement last interaction tracker

### PR DESCRIPTION
**Pull request by Sumit112192**
### :pencil: Description

**Type:** :rocket: `feature`

This continue is based on the work of the previous PR #2650. Also, PR #2714 should be merged first.

### CODE REVIEW

**Comment by Sumit112192** [comment]
@wkerzendorf @andrewfullard Please review.

**Comment by Sumit112192** [comment]
Profiling:
Master Branch
Screenshot from 2024-07-18 14-13-45

ImplementLastInteractionTracker Branch
Screenshot from 2024-07-18 14-14-33

Memory usage has gone down even with the LastInteractionTracker initialization.

File: tardis/transport/montecarlo/single_packet_loop.py [hunk 1]
```diff
@@ -257,9 +257,24 @@ def single_packet_loop(
             )
         else:
             pass
-        if montecarlo_globals.ENABLE_RPACKET_TRACKING:
+        if interaction_type != InteractionType.BOUNDARY:
             rpacket_tracker.track(r_packet)
 
+    # Registering the final boundary interaction.
+    # Only for RPacketTracker
```
Comment (andrewfullard) [inline tardis/transport/montecarlo/single_packet_loop.py:264]: RPacketPlotter then?

File: tardis/transport/montecarlo/single_packet_loop.py [hunk 2]
```diff
@@ -257,9 +257,24 @@ def single_packet_loop(
...
+    if montecarlo_globals.ENABLE_RPACKET_TRACKING:
+        temp_r_packet = RPacket(
```
Comment (andrewfullard) [inline tardis/transport/montecarlo/single_packet_loop.py:266]: rpacket.interaction_type = Boundary
rpacket_tracker.track(r_packet)

Comment (andrewfullard) [reply tardis/transport/montecarlo/single_packet_loop.py:266]: (roughly)

**Review by andrewfullard — APPROVED** [review]

**Pull request merged**